nav2_behaviors/src/spin.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath>
+#include <chrono>
+#include <ctime>
+#include <thread>
+
+#include "nav2_behaviors/spin.hpp"
+#include "tf2/LinearMath/Quaternion.h"
+#include "tf2/LinearMath/Matrix3x3.h"
+
+using nav2_tasks::TaskStatus;
+using namespace std::chrono_literals;
+
+namespace nav2_behaviors
+{
+
+Spin::Spin() : Node("Spin")
+{
+  RCLCPP_INFO(get_logger(), "Initializing the Spin behavior");
+
+  // TODO(orduno) Pull values from param server or robot
+  max_rotational_vel_ = 1.0;
+  min_rotational_vel_ = 0.4;
+  rotational_acc_lim_ = 3.2;
+  goal_tolerance_angle_ = 0.10;
+
+  auto temp_node = std::shared_ptr<rclcpp::Node>(this, [](auto) {});
+
+  robot_ = std::make_unique<nav2_robot::Robot>(temp_node);
+
+  vel_pub_ =
+    this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 1);
+
+  task_server_ = std::make_unique<nav2_tasks::SpinTaskServer>(temp_node, false);
+  task_server_->setExecuteCallback(std::bind(&Spin::spin, this, std::placeholders::_1));
+
+  // Start listening for incoming Spin task requests
+  task_server_->startWorkerThread();
+
+  RCLCPP_INFO(get_logger(), "Initialized the Spin  behavior");
+}
+
+Spin::~Spin()
+{
+  RCLCPP_INFO(get_logger(), "Shutting down the Spin behavior");
+}
+
+nav2_tasks::TaskStatus Spin::spin(const nav2_tasks::SpinCommand::SharedPtr command)
+{
+  RCLCPP_INFO(get_logger(), "Attempting to spin");
+
+  double yaw, pitch, roll;
+  getAnglesFromQuaternion(command->quaternion, yaw, pitch, roll);
+
+  if (roll != 0.0 || pitch != 0.0) {
+    RCLCPP_INFO(get_logger(), "Spinning on Y and X not supported"
+    " , will only spin in Z.");
+  }
+
+  RCLCPP_INFO(get_logger(), "Currently only supported spinning by a fixed amount");
+  TaskStatus status = timedSpin();
+
+  nav2_tasks::SpinResult result;
+  task_server_->setResult(result);
+
+  return status;
+}
+
+nav2_tasks::TaskStatus Spin::timedSpin()
+{
+  auto start_time = std::chrono::system_clock::now();
+  auto time_since_msg = std::chrono::system_clock::now();
+
+  while (true) {
+    if (task_server_->cancelRequested()) {
+      RCLCPP_INFO(get_logger(), "Task cancelled");
+      task_server_->setCanceled();
+      return TaskStatus::CANCELED;
+    }
+
+    // TODO(orduno): For now, spinning is time based
+    auto duration = 4s;
```

Comment:
Could be a parameter in the command message.

Reply:
See previous comment in `back_up.cpp`